runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -417,68 +421,71 @@ public void invalidateServerCaches() {
         });
     }
 
-    /** Force the client cache to be invalidated. */
+    /**
+     * Force the client cache to be invalidated.
+     */
     public void invalidateClientCache() {
         readCache.invalidateAll();
     }
 
     /**
      * Fetch a collection of addresses for insertion into the cache.
      *
-     * @param addresses collection of addresses to read from.
-     * @return A result to be cached
+     * @param addresses    collection of addresses to read from.
+     * @param waitForWrite flag whether wait for write is required or hole fill directly.
+     * @return a map of read addresses.
      */
-    public @Nonnull
-    Map<Long, ILogData> fetchAll(Iterable<Long> addresses) {
+    @Nonnull
+    public Map<Long, ILogData> fetchAll(Iterable<Long> addresses, boolean waitForWrite) {
         Map<Long, ILogData> result = new HashMap<>();
 
         Iterable<List<Long>> batches = Iterables.partition(addresses,
-            runtime.getParameters().getBulkReadSize());
+                runtime.getParameters().getBulkReadSize());
 
         for (List<Long> batch : batches) {
             try {
-                //doesn't handle the case where some address have a different replication mode
+                // Doesn't handle the case where some address have a different replication mode
                 result.putAll(layoutHelper(e -> e.getLayout()
                         .getReplicationMode(batch.iterator().next())
                         .getReplicationProtocol(runtime)
-                        .readAll(e, batch)));
+                        .readAll(e, batch, waitForWrite)));
             } catch (Exception e) {
-                log.error("cacheFetch: Couldn't read addresses {}", batch, e);
-                throw new UnrecoverableCorfuError(
-                    "Unexpected error during cacheFetch", e);
+                log.error("fetchAll: Couldn't read addresses {}", batch, e);
+                throw new UnrecoverableCorfuError("Unexpected error during fetchAll", e);
             }
         }
 
-        for (Long address : result.keySet()) {
-            checkLogData(address, result.get(address));
-        }
+        result.forEach(this::checkLogData);
         return result;
     }
 
     /**
-     * Fetch a range of addresses.
+     * Fetch a closed range of addresses.
      *
-     * @param addresses collection of addresses to read from.
-     * @return A result to be cached
+     * @param lower lower endpoint of the range, included.
+     * @param upper upper endpoint of the range, included.
+     * @return a map of read addresses.
      */
-    public @Nonnull
-    Map<Long, ILogData> fetchRange(Set<Long> addresses) {
-        Map<Long, ILogData> result = layoutHelper(e -> e.getLayout().getReplicationMode(addresses.iterator().next())
+    @Nonnull
+    public Map<Long, ILogData> fetchRange(long lower, long upper, boolean waitForWrite) {
```

Comment:
Why there is no limit on the bulk size for fetchRange() as for fetchAll()?

Reply:
To keep order of result map. FastObjectLoader expects an ordered result.

Reply:
Anyway I'm removing this API to only do sorting in FastObjectLoader, which makes it cleaner.